### ISSUE TITLE
dzil test requires setup configuration

### ISSUE THREAD

**Issue by szabgab** [perlancar/perl-Sah-Schemas-Path#1]
Hi Perlancar,

I was trying to check if I could create a Travis-CI configuration to this module, but I got stuck at this point where it wanted me to configure Dist::Zilla first.

How could this be fixed to enable a CI setup?

```
$ dzil test
[DZ] building distribution under .build/7RO_L8Tnmz for installation
[@Author::PERLANCAR/CopyrightYearFromGit] Setting copyright_year to 2020, 2019, 2018, 2016
[DZ] beginning to build Sah-Schemas-Path
[Sah::Schemas] Adding xt/release/sah-schema.t ...
[DZ] guessing dist's main_module is lib/Sah/Schemas/Path.pm
[DZ] no license data in config, no %Rights stash, couldn't make a good guess at license from Pod; giving up.  Perhaps you need to set up a global config file (dzil setup)?
[DZ] no license data in config, no %Rights stash, couldn't make a good guess at license from Pod; giving up.  Perhaps you need to set up a global config file (dzil setup)? at inline delegation in Dist::Zilla for logger->log_fatal (attribute declared in /home/gabor/perl5/lib/perl5/Dist/Zilla.pm at line 773) line 18.
```

Temporarily I configured Dist::Zilla (without  my PAUSE credentials) and ran the `dzil test` again. This time it failed with this:

```
[@Author::PERLANCAR/Authority] PAUSE credentials not found in "config.ini" or "dist.ini" or "~/.pause"! Please set it or specify an authority for this plugin.
[@Author::PERLANCAR/Authority] PAUSE credentials not found in "config.ini" or "dist.ini" or "~/.pause"! Please set it or specify an authority for this plugin. at inline delegation in Dist::Zilla::Plugin::Authority for logger->log_fatal (attribute declared in /home/gabor/perl5/lib/perl5/Dist/Zilla/Role/Plugin.pm at line 59) line 18.
```

That will probably out-rule a CI, but it will also make it difficult for someone to send a PR without having a PAUSE account.

**Comment by perlancar** [comment]
Hi Gabor,

I've replaced [Authority] plugin with my own version [PERLANCAR::Authority] which, instead of bailing out, uses a placeholder authority when PAUSE credential configuration is not found. Hope it eases CI setup and testing by contributors.

Could you try again with my Dist::Zilla::PluginBundle::Author::PERLANCAR 0.601? Thanks.

**Comment by szabgab** [comment]
I've installed the latest Dist::Zilla::PluginBundle::Author::PERLANCAR noe 0.602 and it is still the same.

**Comment by perlancar** [comment]
Could you show the output? I've just tested from a vanilla perlbrew
installation and I can 'dzil test' without any [%PAUSE] section in dzil
configuration.

BTW, most of my distros can also use 'prove -l' with no problem.

**Comment by szabgab** [comment]
```
[DZ] building distribution under .build/A6r5rnCpM6 for installation
[@Author::PERLANCAR/CopyrightYearFromGit] Setting copyright_year to 2020, 2019, 2018, 2016
[DZ] beginning to build Sah-Schemas-Path
[Sah::Schemas] Adding xt/release/sah-schema.t ...
[DZ] guessing dist's main_module is lib/Sah/Schemas/Path.pm
[DZ] no license data in config, no %Rights stash, couldn't make a good guess at license from Pod; giving up.  Perhaps you need to set up a global config file (dzil setup)?
[DZ] no license data in config, no %Rights stash, couldn't make a good guess at license from Pod; giving up.  Perhaps you need to set up a global config file (dzil setup)? at inline delegation in Dist::Zilla for logger->log_fatal (attribute declared in /home/gabor/perl5/lib/perl5/Dist/Zilla.pm at line 773) line 18.

```

In any case the goal was to set up a Travis-CI for this repo. I got to this point in my attempts: https://travis-ci.org/github/szabgab/perl-Sah-Schemas-Path/jobs/719485309

**Comment by szabgab** [comment]
In a nutshell, don't worry about me. IF it works for you and everyone else then don't waste time trying to fix it for me.
If anything it would be nice if you set up a CI system for this repository.

**Comment by perlancar** [comment]
Hi Gabor,

I've tried adding a minimal `.travis.yml` to one of my distros here: https://github.com/perlancar/perl-Log-ger . I opted to use `prove -l` for now instead of building the distro with dzil. No luck yet though.

https://travis-ci.com/github/perlancar/perl-Log-ger/builds/181510364?utm_medium=notification&utm_source=email

**Comment by szabgab** [comment]
I'd probably ask cpanm to be more verbose so I might be able to see the
error.

**Comment by perlancar** [comment]
Invoking `cpanm` is not in my `.travis.yml`, but I'll read up more later, thanks.

**Comment by szabgab** [comment]
I started to setup GitHub Actions instead of travis, it gives a more verbose output. Does this help locating the issue ?

https://github.com/szabgab/perl-Sah-Schemas-Path/runs/1313153682?check_suite_focus=true

**Comment by perlancar** [comment]
I do have ~/.dzil/config.ini like below. Not sure how to handle this in GitHub Actions or travis.

    [%User]
    name  = perlancar
    email = perlancar@cpan.org

    [%Rights]
    license_class    = Perl_5
    copyright_holder = perlancar@cpan.org

    [%PAUSE]
    username = PERLANCAR
    password = xxxxxxxxxxxxxxx

**Comment by szabgab** [comment]
Forget about the CI systems, I have the same issue on my own computer:

```
$ dzil listdeps
[DZ] no license data in config, no %Rights stash, couldn't make a good guess at license from Pod; giving up.  Perhaps you need to set up a global config file (dzil setup)?
[DZ] no license data in config, no %Rights stash, couldn't make a good guess at license from Pod; giving up.  Perhaps you need to set up a global config file (dzil setup)? at inline delegation in Dist::Zilla for logger->log_fatal (attribute declared in /home/gabor/perl5/lib/perl5/Dist/Zilla.pm at line 773) line 18.
```

I don't have a ~/.dzil/config.ini  I am not sure if requiring such a configuration is a reasonable requirement in the dzil world or not for a contributor. I am not familiar with dzil. Let's say it is. I created a file `~/.dzil/config.ini` with

```
[%User]
name  = perlancar
email = perlancar@cpan.org

[%Rights]
license_class    = Perl_5
copyright_holder = perlancar@cpan.org
```
and ran

```
dzil authordeps --missing | cpanm
```
That did not have to do anything

Then

```
dzil listdeps
```

told me:

```
Pod::Weaver::PluginBundle::Author::PERLANCAR (for section @Author::PERLANCAR) does not appear to be installed
```


Pod::Weaver::PluginBundle::Author::PERLANCAR (for section @Author::PERLANCAR) does not appear to be installed

**Comment by szabgab** [comment]
I made some progress but  I had to install several dependencies "manually"

           cpanm --notest Pod::Weaver::PluginBundle::Author::PERLANCAR
            cpanm --notest Pod::Weaver::Plugin::Data::Sah::Coerce
            cpanm --notest Pod::Weaver::Plugin::Sah::Schemas

but then when I ran

dzil test

I got the following error:

pmversions file '/github/home/pmversions.ini' does not exist at
/usr/local/lib/perl5/site_perl/5.30.3/PMVersions/Util.pm line 25.

Any idea what should I do with this?

Full output:
https://github.com/szabgab/perl-Sah-Schemas-Path/runs/1369806186?check_suite_focus=true

**Comment by perlancar** [comment]
I have released PMVersions-Util-0.002 which will just warn instead of die when pmversions.ini does not exist.

**Comment by szabgab** [comment]
Now it requires  Test::Perl::Critic to be installed for t/author-critic.t  Was this test added on-the-fly by Dist::Zilla?
https://github.com/szabgab/perl-Sah-Schemas-Path/runs/1384597004?check_suite_focus=true

After installing Test::Perl::Critic it reported all kinds of issues.

https://github.com/szabgab/perl-Sah-Schemas-Path/runs/1384638911?check_suite_focus=true

**Comment by perlancar** [comment]
Yes, a test script that uses Test::Perl::Critic is added by d Dist::Zilla plugin, namely Dist::Zilla::Plugin::Test::Perl::Critic::Subset. It's an author-only test and depends on perl critic configuration.

I just realize that "dzil test" runs author tests by default. I think for CI it should be "dzil test --no-author". Could you use that instead?

**Comment by szabgab** [comment]
That way it works and I am going to send you the pull-request, but I you'll get a much better value from the CI if we figure out how to run all the tests, including the author tests and release-tests if you have them.

For that either you can have your perl critic configuration included in the repo (but excluded from the distribution) or, if you have it some other public place the CI could copy it from there.
